File: lib/src/domain/scoring.dart
import 'models.dart';

class ActualResults {
  const ActualResults({
    required this.knockoutWinnersBySlot,
    this.groupPlacements,
    this.finalChampionScore,
    this.finalRunnerUpScore,
  });

  final OfficialGroupPlacements? groupPlacements;
  final Map<String, String> knockoutWinnersBySlot;
  final int? finalChampionScore;
  final int? finalRunnerUpScore;
}

class ScoreBreakdown {
  const ScoreBreakdown({
    required this.groupScore,
    required this.knockoutScore,
    required this.totalScore,
    required this.tiebreakerDistance,
  });

  final int groupScore;
  final int knockoutScore;
  final int totalScore;
  final int tiebreakerDistance;
}

class BracketScorer {
  const BracketScorer();

  /// Points when a picked team finishes in the group's top three.
  static const groupTopThreePoints = 1;

  /// Extra points when the team finishes in the exact position you picked.
  static const groupExactPlacementBonus = 2;

  /// Knockout ladder: each round doubles (R32 = 1 pt, Final = 16 pts).
  static const knockoutPointsByStage = {
    TournamentStage.roundOf32: 1,
    TournamentStage.roundOf16: 2,
    TournamentStage.quarterfinal: 4,
    TournamentStage.semifinal: 8,
    TournamentStage.finalMatch: 16,
  };

  static int knockoutPointsForStage(TournamentStage stage) {
    return knockoutPointsByStage[stage] ?? 0;
  }

  ScoreBreakdown score({
    required Bracket bracket,
    required ActualResults results,
    required int pointsPerCorrectPick,
  }) {
    final groupScore = _scoreGroupPlacements(
      bracket.groupPicks,
      results.groupPlacements,
    );
    final knockoutScore = bracket.knockoutPicks.fold<int>(0, (score, pick) {
      final actualWinner = results.knockoutWinnersBySlot[pick.slotId];
      if (actualWinner == null || actualWinner != pick.winnerCountryId) {
        return score;
      }
      return score + knockoutPointsForStage(pick.stage);
    });

    final tiebreakerDistance = _tiebreakerDistance(bracket, results);
    return ScoreBreakdown(
      groupScore: groupScore,
      knockoutScore: knockoutScore,
      totalScore: groupScore + knockoutScore,
      tiebreakerDistance: tiebreakerDistance,
    );
  }

  int _scoreGroupPlacements(
    List<GroupPick> predictedPicks,
    OfficialGroupPlacements? actualPlacements,
  ) {
    if (actualPlacements == null) {
      return 0;
    }

    var score = 0;
    for (final predicted in predictedPicks) {
      final actual = _officialGroupPick(actualPlacements, predicted.groupId);
      if (actual == null) {
        continue;
      }
      score += _scorePredictedSlot(
        predictedCountryId: predicted.firstCountryId,
        predictedRank: 1,
        actual: actual,
      );
      score += _scorePredictedSlot(
        predictedCountryId: predicted.secondCountryId,
        predictedRank: 2,
        actual: actual,
      );
      score += _scorePredictedSlot(
        predictedCountryId: predicted.thirdCountryId ?? '',
        predictedRank: 3,
        actual: actual,
      );
    }
    return score;
  }

  int _scorePredictedSlot({
    required String predictedCountryId,
    required int predictedRank,
    required GroupPick actual,
  }) {
    if (predictedCountryId.isEmpty) {
      return 0;
    }
    if (!_topThreeCountryIds(actual).contains(predictedCountryId)) {
      return 0;
    }
    if (predictedCountryId == _countryIdAtRank(actual, predictedRank)) {
      return groupTopThreePoints + groupExactPlacementBonus;
    }
    return groupTopThreePoints;
  }

  GroupPick? _officialGroupPick(
    OfficialGroupPlacements placements,
    String groupId,
  ) {
    for (final pick in placements.groupPicks) {
      if (pick.groupId == groupId) {
        return pick;
      }
    }
    return null;
  }

  Set<String> _topThreeCountryIds(GroupPick pick) {
    return {
      pick.firstCountryId,
      pick.secondCountryId,
      if (pick.thirdCountryId != null && pick.thirdCountryId!.isNotEmpty)
        pick.thirdCountryId!,
    };
  }

  String _countryIdAtRank(GroupPick pick, int rank) {
    switch (rank) {
      case 1:
        return pick.firstCountryId;
      case 2:
        return pick.secondCountryId;
      case 3:
        return pick.thirdCountryId ?? '';
      default:
        return '';
    }
  }

  int _tiebreakerDistance(Bracket bracket, ActualResults results) {
    final championScore = results.finalChampionScore;
    final runnerUpScore = results.finalRunnerUpScore;
    if (championScore == null || runnerUpScore == null) {
      return 0;
    }
    return bracket.finalScoreTiebreaker.distanceFrom(
      championScore,
      runnerUpScore,
    );
  }
}

File: functions/src/scoring.ts
import {GroupPick, OfficialGroupPlacements} from "./group-placements";

export type KnockoutPick = {
  slotId: string;
  stage: string;
  winnerCountryId: string;
};

export type FinalScoreTiebreaker = {
  championScore?: number;
  runnerUpScore?: number;
};

export type BracketRecord = {
  status?: string;
  groupPicks?: GroupPick[];
  knockoutPicks?: KnockoutPick[];
  finalScoreTiebreaker?: FinalScoreTiebreaker;
};

export type OfficialResultsRecord = {
  groupPlacements?: OfficialGroupPlacements | null;
  knockoutWinnersBySlot?: Record<string, string>;
  finalChampionScore?: number | null;
  finalRunnerUpScore?: number | null;
};

export type ScoreBreakdown = {
  groupScore: number;
  knockoutScore: number;
  totalScore: number;
  tiebreakerDistance: number;
};

const GROUP_TOP_THREE_POINTS = 1;
const GROUP_EXACT_PLACEMENT_BONUS = 2;

const KNOCKOUT_POINTS_BY_STAGE: Record<string, number> = {
  roundOf32: 1,
  roundOf16: 2,
  quarterfinal: 4,
  semifinal: 8,
  finalMatch: 16,
};

export function knockoutPointsForStage(stage: string): number {
  return KNOCKOUT_POINTS_BY_STAGE[stage] ?? 0;
}

export function scoreBracket(
  bracket: BracketRecord,
  officialResults: OfficialResultsRecord,
): ScoreBreakdown {
  const groupScore = scoreGroupPlacements(
    bracket.groupPicks ?? [],
    officialResults.groupPlacements,
  );
  const knockoutScore = (bracket.knockoutPicks ?? []).reduce((score, pick) => {
    const actualWinner = officialResults.knockoutWinnersBySlot?.[pick.slotId];
    if (actualWinner == null || actualWinner !== pick.winnerCountryId) {
      return score;
    }
    return score + knockoutPointsForStage(pick.stage);
  }, 0);
  const tiebreakerDistance = tiebreakerDistanceFor(
    bracket.finalScoreTiebreaker,
    officialResults.finalChampionScore,
    officialResults.finalRunnerUpScore,
  );

  return {
    groupScore,
    knockoutScore,
    totalScore: groupScore + knockoutScore,
    tiebreakerDistance,
  };
}

export function scoreGroupPlacements(
  predictedPicks: GroupPick[],
  actualPlacements: OfficialGroupPlacements | null | undefined,
): number {
  if (actualPlacements?.groupPicks?.length == null) {
    return 0;
  }

  const actualByGroup = new Map(
    actualPlacements.groupPicks.map((pick) => [pick.groupId, pick]),
  );

  let score = 0;
  for (const predicted of predictedPicks) {
    const actual = actualByGroup.get(predicted.groupId);
    if (actual == null) {
      continue;
    }
    score += scorePredictedSlot(predicted.firstCountryId, 1, actual);
    score += scorePredictedSlot(predicted.secondCountryId, 2, actual);
    score += scorePredictedSlot(predicted.thirdCountryId ?? "", 3, actual);
  }
  return score;
}

function scorePredictedSlot(
  predictedCountryId: string,
  predictedRank: number,
  actual: GroupPick,
): number {
  if (predictedCountryId.length === 0) {
    return 0;
  }
  if (!topThreeCountryIds(actual).has(predictedCountryId)) {
    return 0;
  }
  if (predictedCountryId === countryIdAtRank(actual, predictedRank)) {
    return GROUP_TOP_THREE_POINTS + GROUP_EXACT_PLACEMENT_BONUS;
  }
  return GROUP_TOP_THREE_POINTS;
}

function topThreeCountryIds(pick: GroupPick): Set<string> {
  return new Set(
    [pick.firstCountryId, pick.secondCountryId, pick.thirdCountryId].filter(
      (countryId): countryId is string =>
        countryId != null && countryId.length > 0,
    ),
  );
}

function countryIdAtRank(pick: GroupPick, rank: number): string {
  switch (rank) {
    case 1:
      return pick.firstCountryId;
    case 2:
      return pick.secondCountryId;
    case 3:
      return pick.thirdCountryId ?? "";
    default:
      return "";
  }
}

function tiebreakerDistanceFor(
  tiebreaker: FinalScoreTiebreaker | undefined,
  championScore: number | null | undefined,
  runnerUpScore: number | null | undefined,
): number {
  if (championScore == null || runnerUpScore == null) {
    return 0;
  }
  return (
    Math.abs((tiebreaker?.championScore ?? 0) - championScore) +
    Math.abs((tiebreaker?.runnerUpScore ?? 0) - runnerUpScore)
  );
}
